src/plexil-adapter/OwAdapter.cpp

```diff
@@ -78,19 +79,31 @@ static void ack_sent (Command* cmd, AdapterExecInterface* intf)
   ack_command (cmd, COMMAND_SENT_TO_SYSTEM, intf);
 }
 
-static void command_status_callback (int id, bool success)
+static void send_ack_once(Command* cmd, AdapterExecInterface* intf, bool* sent_flag)
 {
-  pair<Command*, AdapterExecInterface*> cmd;
-
-  try {
-    cmd = CommandRegistry.at(id);
+  std::lock_guard<std::mutex> g(g_shared_mutex);
+  if (! (*sent_flag) )
+  {
+    ack_sent (cmd, intf);
+    *sent_flag = true;
   }
-  catch (const std::out_of_range& oor) {
-    ROS_ERROR("command_status_callback: no command registered under id %d", id);
+}
+
+static void command_status_callback (int id, bool success)
+{
+  auto it = CommandRegistry.find(id);
+  if (it == CommandRegistry.end())
+  {
+    ROS_ERROR_STREAM("command_status_callback: no command registered under id" << id);
     return;
   }
-  if (success) ack_success (cmd.first, cmd.second);
-  else ack_failure (cmd.first, cmd.second);
+
+  std::unique_ptr<CommandRecord>& cr = it->second;
+  Command* cmd = std::get<0>(*cr);
+  AdapterExecInterface* intf = std::get<1>(*cr);
+  send_ack_once(cmd, intf, &std::get<2>(*cr));
```

Comment:
@mikedalal I had a question for you here. If callback already back before we send the acknowledgment, can't we just skip sending the ack in this case? For example the log_* methods send success signals without ack signal.

Reply:
Yes, I'm pretty sure the ack can be skipped if we have a return status (success, failure).

Reply:
Do you think that would be better approach?

Reply:
Stand by, I'm reviewing the PLEXIL interface documentation...

Reply:
As far as I can tell from the Plexil docs, the equivalent of ack_sent() is not needed at all if ack_success() or ack_failure() can be called first.   I've asked Chuck Fry, the PLEXIL systems expert.

Reply:
More importantly, I actually don't think that doing ack_sent() AFTER success/failure would cause a problem unless the Plexil code had logic that dealt with these command handles, which it doesn't.  I've asked Chuck about this as well.

Reply:
I think this might save some cycles if we do so, I will try it out first before I commit any changes to this regard

Reply:
I didn't dig into why this sending these commands out of order seems to cause a problem for PLEXIL (it may not be a PLEXIL issue but I didn't look further into the problem).

Reply:
It's still a great catch!  But to be honest, I don't see why flipped order would cause the stall problem, or any problem.  The redundant ack should just be ignored.

Reply:
I changed the code such that we skip sending _ack_ when success/failure already returned.
I did test it multiple times and didn't get any stalls

Reply:
This sounds fine.  Chuck confirmed my guess, that ack_sent() after ack_success/failure should have no effect as far as the Plexil executive is concerned.  In fact the ack_sent() handle has no treatment by the executive, only by plans if so coded.   

However, note that these functions wrap an essential call, notifyOfExternalEvent.  This is absolutely necessary to "wake up" the exec.   My first suspicion of the stalling problem is that this was not getting called sometimes, i.e. none of ack_sent, ack_success, or ack_failure were called after starting the pan/tilt.   My debug settings would catch this, yet alas the stall hasn't happened since I put them in!